service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/transport/amqps/AmqpConnectionHandler.java

```diff
@@ -0,0 +1,185 @@
+        }
+        else
+        {
+            this.hostName = hostName + AMQPS_PORT;
+        }
+
+        this.userName = userName;
+        this.sasToken = sasToken;
+    }
+
+    @Override
+    public void onReactorInit(Event event)
+    {
+        event.getReactor().connection(this);
+    }
+
+    /**
+     * Event handler for the connection bound event
+     * @param event The proton event object
+     */
+    @Override
+    public void onConnectionBound(Event event)
+    {
+        Transport transport = event.getConnection().getTransport();
+        if (transport != null)
+        {
+            if (this.iotHubServiceClientProtocol == IotHubServiceClientProtocol.AMQPS_WS)
+            {
+                WebSocketImpl webSocket = new WebSocketImpl();
+                webSocket.configure(this.webSocketHostName, WEBSOCKET_PATH, 0, WEBSOCKET_SUB_PROTOCOL, null, null);
+                ((TransportInternal)transport).addTransportLayer(webSocket);
+            }
+            Sasl sasl = transport.sasl();
+            sasl.plain(this.userName, this.sasToken);
+
+            SslDomain domain = makeDomain(SslDomain.Mode.CLIENT);
+            domain.setPeerAuthentication(SslDomain.VerifyMode.VERIFY_PEER);
+            Ssl ssl = transport.ssl(domain);
+        }
+    }
+
+    @Override
+    public void onLinkRemoteOpen(Event event)
+    {
+        super.onLinkRemoteOpen(event);
+        this.linkOpenedRemotely = true;
+    }
+
+    @Override
+    public void onSessionRemoteOpen(Event event)
+    {
+        super.onSessionRemoteOpen(event);
+        this.sessionOpenedRemotely = true;
+    }
+
+    @Override
+    public void onConnectionRemoteOpen(Event event)
+    {
+        super.onConnectionRemoteOpen(event);
+        this.connectionOpenedRemotely = true;
+    }
+
+    /**
+     * If an exception was encountered while opening the AMQP connection, this function shall throw that saved exception
+     * @throws IOException if an exception was encountered while openinging the AMQP connection. The encountered
+     * exception will be the inner exception
+     */
+    protected void verifyConnectionOpened() throws IOException
+    {
+        if (this.protonJExceptionParser != null)
+        {
+            throw new IOException("Encountered exception during amqp connection: " + protonJExceptionParser.getError() + " with description " + protonJExceptionParser.getErrorDescription());
+        }
+
+        if (this.savedException != null)
+        {
+            throw new IOException("Connection failed to be established", this.savedException);
+        }
+
+        if (!this.connectionOpenedRemotely)
+        {
+            throw new IOException("Amqp Connection failed to open");
+        }
+
+        if (!this.sessionOpenedRemotely)
+        {
+            throw new IOException("Amqp Session failed to open");
+        }
+
+        if (!this.linkOpenedRemotely)
+        {
+            throw new IOException("Amqp link never opened");
+        }
```

Comment:
This doesn't really give any context about why the connection/session/link wasn't opened. I am not sure if simply throwing an IOException here without an inner exception is of any value here. I would expect that if connection/session/links were not opened successfully, then there would be a savedException saved. 
These lines should be simplified accordingly.

Reply:
The problem here is that there isn't necessarily an exception thrown by opening the connection. If we simply time out while opening the connection, proton-j may not give us any exception to include here. If protonJExceptionParser and savedException are both null, then we have no extra context to give here. All we know is that the connection/session/link did not open remotely